### PR TITLE
Add SAM3 Canvas Mode for Cross Image detection 

### PR DESCRIPTION
**Pull request by rajeshgangireddy**
# Pull Request

Adds "Canvas Mode" as an alternate to the VisualExemplar Mode for cross image detection for SAM3.
VisualExemplar mode was not working properly (low accuracy) when the text category name was not provided or was not named properly.  For users, this would mean that they have to use proper category names that matched the object they are trying to segment. Moreoever, with multiple categories, VisualExemplar mode failed. 

In this PR, introduce [FSS-SAM3 paper's ](https://arxiv.org/abs/2604.05433) idea of using a unified canvas mode with reference and target image in the same "canvas". This means, either using the whole reference image or a crop (default) of the annotated reference image to be 30% of the canvas and the remaining space of the canvas to be the target image.  

This mode works better than visual exemplar with and without category name supplied - so it's upto application whether to pass this info or not (maybe we can add a checkbox for "Use labels info"   in UI ?).  Also tested with multiple categories. Need to provide category ids for multple categories. 

## Benchmarks

Dataset | Category | VE(no text) | VE+Category Info | Canvas(no Category Info) | Canvas+Category Info | Improvement (No Category Info) | Delta (With  Category Info)
-- | -- | -- | -- | -- | -- | -- | --
LVIS | cupcake | 49.9 | 60.0 | 56.9 | 61.8 | 7.0 | 1.8
LVIS | strawberry | 51.7 | 66.0 | 65.4 | 76.2 | 13.7 | 10.2
LVIS | mushroom | 7.4 | 26.5 | 22.5 | 37.5 | 15.1 | 11.0
LVIS | sheep | 81.4 | 86.1 | 86.3 | 85.1 | 4.9 | -1.0
LVIS | giraffe | 4.7 | 81.3 | 80.0 | 82.4 | 75.3 | 1.1
PerSeg | teddybear | 0.0 | 83.3 | 100.0 | 100.0 | 100.0 | 16.7
PerSeg | elephant | 0.0 | 50.0 | 100.0 | 100.0 | 100.0 | 50.0
PerSeg | duck_toy | 0.0 | 100.0 | 100.0 | 100.0 | 100.0 | 0.0
PerSeg | candle | 36.4 | 100.0 | 100.0 | 100.0 | 63.6 | 0.0
PerSeg | cat | 100.0 | 100.0 | 100.0 | 100.0 | 0.0 | 0.0
  |   |   |   |   |   |   |  
Average | Average | 33.2 | 75.3 | 81.1 | 84.3 | 48.0 | 9.0

### Example 1 : Single Category with and without category information passed. 
image

image

### Example 2 : Multiple categories with and without category information passed. 
Note that for motorcyle, when the catrgory name is passed, the masks segmented only contains motocyle and excludes rider. 
image

## Usage 

```python
from instantlearn.data import Sample
from instantlearn.models.sam3 import SAM3, CanvasConfig, Sam3PromptMode

model = SAM3(
    prompt_mode=Sam3PromptMode.CANVAS,
    canvas_config=CanvasConfig(split_ratio=0.3, crop_padding=2.0), # Optional, defaults are already tuned to get best results for most of cases
    confidence_threshold=0.3,
)

# Visual-only (no text)
ref = Sample(image_path="ref.jpg", bboxes=np.array([[x1, y1, x2, y2]]))
model.fit(ref)
predictions = model.predict([Sample(image_path="target.jpg")])

# With text category
ref = Sample(image_path="ref.jpg", bboxes=bbox, categories=["SUV"], category_ids=np.array([0]))
model.fit(ref)
predictions = model.predict([Sample(image_path="target.jpg")])

# Multi-category
ref_a = Sample(image_path="ref.jpg", bboxes=bbox_a, categories=["SUV"], category_ids=np.array([0]))
ref_b = Sample(image_path="ref.jpg", bboxes=bbox_b, categories=["Motorcycle"], category_ids=np.array([1]))
model.fit([ref_a, ref_b])
predictions = model.predict([Sample(image_path="target.jpg")])
```

## Type of Change

- [x] ✨ `feat` - New feature
- [ ] 🐞 `fix` - Bug fix
- [ ] 📚 `docs` - Documentation
- [ ] ♻️ `refactor` - Code refactoring
- [ ] 🧪 `test` - Tests
- [ ] 🔧 `chore` - Maintenance

## Breaking Changes

---

### CODE REVIEW

**Review by Daankrol — APPROVED** [review]
Good upgrade!
